### PR TITLE
OO-DOCI driver as a pybinding (resolves #347).

### LINKED ISSUE #347: Remove all C++ drivers in favor of Python?
Since we've currently (see PR #334) implemented FCI and Hubbard as `QCMethod`, it would be nice to also implement OO-DOCI as one and remove the `drivers` folder altogether.

### PR DESCRIPTION
**Pull request by tmhuysen**
This feature adds the final driver as a pybinding, and includes tests.

- [x]  OO-doci driver
- [x]  Test for a single pybinding to evaluate the format

Should the QCM classes made for the bindings have tests or can they be implicitly tested through bindings?

### CODE REVIEW

File: gqcpy/QCMethod/FCI.cpp [hunk 1]
```diff
@@ -30,8 +30,8 @@ namespace gqcpy {
 
 void bindQCMethodFCI(py::module& module) {
     py::class_<GQCP::QCMethod::FCI>(module, "FCI", "Construct and solve the FCI Hamiltonian.")
-        .def(py::init<const std::string, const std::string, const size_t, const size_t>(), py::arg("xyz_filename"), py::arg("basis_set"), py::arg("num_alpha"), py::arg("num_beta"))
-        .def("solve", &GQCP::QCMethod::FCI::solve, "Solve the eigenvalue equations such that the lowest energy and corresponding eigenvector become available. ")
+        .def(py::init<const std::string&, const std::string&, const size_t, const size_t>(), py::arg("xyz_filename"), py::arg("basis_set"), py::arg("num_alpha"), py::arg("num_beta"))
+        .def("solve", &GQCP::QCMethod::FCI::solve, "Solve the eigenvalue equations such that the lowest energy and corresponding eigenvector becomes available. ")
```
Comment (lelemmen) [inline gqcpy/QCMethod/FCI.cpp]:
```suggestion
        .def("solve", &GQCP::QCMethod::FCI::solve, "Solve the eigenvalue equations such that the lowest energy and corresponding eigenvector become available. ")
```

energy and ... eigenvector **become**

File: gqcpy/tests/hubbard_binding_test.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import unittest
+import numpy as np
+
+# force the local gqcpy to be imported
```
Comment (lelemmen) [inline gqcpy/tests/hubbard_binding_test.py]:
```suggestion
# Force the local gqcpy to be imported
```

File: gqcpy/tests/hubbard_binding_test.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import sys
+sys.path.insert(0, '../')
```
Comment (lelemmen) [inline gqcpy/tests/hubbard_binding_test.py]: Is this still needed?

File: gqcpy/tests/hubbard_binding_test.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+# import our `pybind11`-based extension module
```
Comment (lelemmen) [inline gqcpy/tests/hubbard_binding_test.py]:
```suggestion
```

File: gqcpy/tests/hubbard_binding_test.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import gqcpy
+
+
+class HubbardQCM(unittest.TestCase):
+
+    ''' iniates variables to be used by tests '''
```
Comment (lelemmen) [inline gqcpy/tests/hubbard_binding_test.py]:
```suggestion
    ''' Iniates variables to be used by tests '''
```

Comment (lelemmen) [reply gqcpy/tests/hubbard_binding_test.py]: Docstrings should be in triple quotes and below the Python definition.

File: gqcpy/tests/hubbard_binding_test.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        self.energy1 = energies[0]
+        self.energy2 = energies[1]
+        rdms = self.hubbardQCM.get_one_rdms()
+        self.rdm1 = rdms[0]
+        self.rdm2 = rdms[1]
+
+    def tearDown(self):
+        pass
+
+        ''' compare energies with reference '''
```
Comment (lelemmen) [inline gqcpy/tests/hubbard_binding_test.py]: Docstrings should be in triple quotes and below the Python definition.

File: gqcpy/tests/hubbard_binding_test.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    def test_energies(self):
+        self.assertAlmostEqual(self.energy1, self.ref_energy1)
+        self.assertAlmostEqual(self.energy2, self.ref_energy2)
+
+        ''' compare RDMs with reference '''
```
Comment (lelemmen) [inline gqcpy/tests/hubbard_binding_test.py]:
```suggestion
        ''' compare RDMs with reference '''
```

File: src/QCMethod/CMakeLists.txt [hunk 8]
```diff
@@ -1,5 +1,6 @@
 target_sources(gqcp
         PRIVATE
+            DOCINewtonOrbitalOptimizer
```
Comment (lelemmen) [inline src/QCMethod/CMakeLists.txt]: I guess this needs a `.cpp`-extension?

File: tests/CMakeLists.txt [hunk 9]
```diff
@@ -40,4 +41,4 @@ foreach(TEST_SOURCE ${test_target_sources})
     add_test(NAME ${TEST_NAME} COMMAND ${TEST_NAME}
             WORKING_DIRECTORY ${CMAKE_BINARY_DIR})  # the working directory is the out-of-source build directory
 
-endforeach()
+endforeach()
```
Comment (lelemmen) [inline tests/CMakeLists.txt]: Add new line at eof

File: tests/QCMethod/Hubbard_QCM_test.cpp [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    auto const& hubbard_eigenpairs = hubbard_solver.get_eigenpairs();
+
+    for (size_t i = 0; i < nos; i++) {
+            BOOST_CHECK(std::abs(hubbard_qcm_energies[i] - (hubbard_eigenpairs[i].get_eigenvalue())) < 1.0e-06);
+            rdm_calc.set_coefficients(hubbard_eigenpairs[i].get_eigenvector());
+            auto const& one_rdm = rdm_calc.calculate1RDMs().one_rdm;
+            BOOST_CHECK(one_rdm.isApprox(hubbard_qcm_1rdms[i], 1.0e-06));
+    }
+}
+
```
Comment (lelemmen) [inline tests/QCMethod/Hubbard_QCM_test.cpp]:
```suggestion
```

File: tests/QCMethod/Hubbard_QCM_test.cpp [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+#include "QCMethod/Hubbard.hpp"
+#include "HamiltonianBuilder/Hubbard.hpp"
+#include "CISolver/CISolver.hpp"
+#include "RDM/RDMCalculator.hpp"
+
+
+BOOST_AUTO_TEST_CASE ( Hubbard_QCM ) {
+
+    // Check if the Hubbard module and the Hubbard QCM produce the same results
+    // retrieve a suitable csline
```
Comment (lelemmen) [inline tests/QCMethod/Hubbard_QCM_test.cpp]:
```suggestion
    // Retrieve a suitable csline
```

File: gqcpy/tests/hubbard_binding_test.py (hunk 3 above)
Comment (tmhuysen) [reply gqcpy/tests/hubbard_binding_test.py]: yes

**Pull request merged**